libnet/ipams/default: introduce a linear allocator

akerouanton opened the pull request:
- Close https://github.com/moby/moby/issues/40275
- Close https://github.com/moby/moby/issues/42801
- Close https://github.com/moby/libnetwork/issues/2612
- Supersede / Close https://github.com/moby/moby/pull/43033
- Stacked on top of https://github.com/moby/moby/pull/47727

**- What I did**

This previous allocator was subnetting address pools eagerly when the daemon started, and would then just iterate over that list whenever RequestPool was called. This was leading to high memory usage whenever IPv6 pools were configured with a target subnet size too different from the pools prefix size.

For instance: pool = fd00::/8, target size = /64 -- 2 ^ (64-8) subnets would be generated upfront. This would take approx. 9 * 10^18 bits -- way too much for any human computer in 2024.

Another noteworthy issue, the previous implementation was allocating a subnet, and then in another layer was checking whether the allocation was conflicting with some 'reserved networks'. If so, the allocation would be retried, etc... To make it worse, 'reserved networks' would be recomputed on every iteration. This is totally ineffective as there could be 'reserved networks' that fully overlap a given address pool (or many!).

To fix this issue, a new field `Exclude` is added to `RequestPool`. It's up to each driver to take it into account. Since we don't know whether this retry loop is useful for some remote IPAM driver, it's reimplemented bug-for-bug directly in the remote driver.

The new allocator uses a linear-search algorithm. It takes advantage of all lists (predefined pools, allocated subnets and reserved networks) being sorted and logically combines 'allocated' and 'reserved' through a 'double cursor' to iterate on both lists at the same time while preserving the total order. At the same time, it iterates over 'predefined' pools and looks for the first empty space that would be a good fit.

Currently, the size of the allocated subnet is still dictated by each 'predefined' pools. We should consider hardcoding that size instead, and let users specify what subnet size they want. This wasn't possible before as the subnets were generated upfront. This new allocator should be able to deal with this easily.

The method used for static allocation has been updated to make sure the ascending order of 'allocated' is preserved. It's bug-for-bug compatible with the previous implementation.

One consequence of this new algorithm is that we don't keep track of where the last allocation happened, we just allocate the first free subnet we find.

Before:

- Allocate: 10.0.1.0/24, 10.0.2.0/24 ; Deallocate: 10.0.1.0/24 ; Allocate 10.0.3.0/24.

Now, the 3rd allocation would yield 10.0.1.0/24 once again.

As it doesn't change the semantics of the allocator, there's no reason to worry about that.

Finally, about 'reserved networks'. The heuristics we use are now properly documented. It was discovered that we don't check routes for IPv6 allocations -- this can't be changed because there's no such thing as on-link routes for IPv6.

(Kudos to Rob Murray for coming up with the linear-search idea.)

**- How to verify it**

CI -- a bunch of tests have been added, some have been rewritten.

Or manually by creating, deleting and re-creating networks.

**- Description for the changelog**

```markdown changelog
- Introduce a new subnet allocator that can deal with IPv6 address pools of any size
```

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://cincinnatizoo.org/wp-content/uploads/2012/04/Red-Panda-Image-for-Endangered-Excursion-scaled.jpg)

akerouanton committed:
libnet/i/defaultipam: introduce a linear allocator

The previous allocator was subnetting address pools eagerly
when the daemon started, and would then just iterate over that
list whenever RequestPool was called. This was leading to high
memory usage whenever IPv6 pools were configured with a target
subnet size too different from the pools prefix size.

For instance: pool = fd00::/8, target size = /64 -- 2 ^ (64-8)
subnets would be generated upfront. This would take approx.
9 * 10^18 bits -- way too much for any human computer in 2024.

Another noteworthy issue, the previous implementation was allocating
a subnet, and then in another layer was checking whether the
allocation was conflicting with some 'reserved networks'. If so,
the allocation would be retried, etc... To make it worse, 'reserved
networks' would be recomputed on every iteration. This is totally
ineffective as there could be 'reserved networks' that fully overlap
a given address pool (or many!).

To fix this issue, a new field `Exclude` is added to `RequestPool`.
It's up to each driver to take it into account. Since we don't know
whether this retry loop is useful for some remote IPAM driver, it's
reimplemented bug-for-bug directly in the remote driver.

The new allocator uses a linear-search algorithm. It takes advantage
of all lists (predefined pools, allocated subnets and reserved
networks) being sorted and logically combines 'allocated' and
'reserved' through a 'double cursor' to iterate on both lists at the
same time while preserving the total order. At the same time, it
iterates over 'predefined' pools and looks for the first empty space
that would be a good fit.

Currently, the size of the allocated subnet is still dictated by
each 'predefined' pools. We should consider hardcoding that size
instead, and let users specify what subnet size they want. This
wasn't possible before as the subnets were generated upfront. This
new allocator should be able to deal with this easily.

The method used for static allocation has been updated to make sure
the ascending order of 'allocated' is preserved. It's bug-for-bug
compatible with the previous implementation.

One consequence of this new algorithm is that we don't keep track
of where the last allocation happened, we just allocate the first
free subnet we find.

Before:

- Allocate: 10.0.1.0/24, 10.0.2.0/24 ; Deallocate: 10.0.1.0/24 ;
Allocate 10.0.3.0/24.

Now, the 3rd allocation would yield 10.0.1.0/24 once again.

As it doesn't change the semantics of the allocator, there's no
reason to worry about that.

Finally, about 'reserved networks'. The heuristics we use are
now properly documented. It was discovered that we don't check
routes for IPv6 allocations -- this can't be changed because
there's no such thing as on-link routes for IPv6.

(Kudos to Rob Murray for coming up with the linear-search idea.)

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

corhere commented on libnetwork/internal/netiputil/netiputil.go:
```diff
@@ -59,3 +59,32 @@ func AddrPortFromNet(addr net.Addr) netip.AddrPort {
+
+// LastAddr returns the last address of prefix 'p'.
+func LastAddr(p netip.Prefix) netip.Addr {
+	return ipbits.Add(ipbits.Sub(p.Addr(), 1, 0), 1, uint(p.Addr().BitLen()-p.Bits()))
```
```suggestion
	return ipbits.Add(p.Addr().Prev(), 1, uint(p.Addr().BitLen()-p.Bits()))
```

robmry reviewed (approved):
LGTM!

akerouanton commented:
@corhere Please make sure this won't break Swarm in any way (eg. when a new leader is elected, etc...).

akerouanton committed:
libnet/i/defaultipam: Unmap IPv4-mapped IPv6 addrs

This ensures such address pools are part of the IPv4 address space.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet/i/defaultipam: improve address pools validation

Nothing was validating whether address pools' `base` prefix
were larger than the target subnet `size` they're associated to. As
such invalid address pools would yield no subnet, the error could go
unnoticed.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

corhere commented on libnetwork/ipams/defaultipam/address_space.go:
```diff
@@ -2,29 +2,54 @@ package defaultipam
+	// one), otherwise the dynamic allocator might consider a predefined
+	// network is fully overlapped, go to the next one, which is a subnet of
+	// the previous, and allocate from it.
+	var last *ipamutils.NetworkToSplit
+	var discarded int
+	for i, imax := 0, len(predefined); i < imax; i++ {
+		p := predefined[i-discarded]
+		if last != nil && last.Overlaps(p.Base) {
+			predefined = slices.Delete(predefined, i-discarded, i-discarded+1)
+			discarded++
+			continue
+		}
+		last = p
+	}
```
Since the `slices` package is already being used, may as well take full advantage.
```suggestion
	predefined = slices.CompactFunc(predefined, func(last, p *ipamutils.NetworkToSplit) bool {
		return last.Overlaps(p.Base)
	})
```

akerouanton replied:
`slices.CompactFunc` works a bit differently. It expects a strict equality as it doesn't compare to the last non-duplicate found, but to 'current-1'. If you have the following subnets:

1. `10.0.0.0/8`
2. `10.0.0.0/16`
3. `10.10.0.0/16`

It tries to compare s1 == s2, and then s2 == s3. That's not what we want.

corhere replied:
That's too bad. I don't like that `slices.Delete` is being called in a loop:
> Delete is O(len(s)-i)

Therefore the worst case time complexity is O(N^2).

[It is doable in O(len(predefined)) time.](https://go.dev/wiki/SliceTricks#in-place-deduplicate-comparable)

akerouanton replied:
Yeah, at first I thought `predefined` wouldn't be filled with that many entries to really matter. But better safe than sorry, I guess. That's now fixed.

corhere reviewed (commented):
Review WIP; I haven't even finished with address_space.go. I'll be back tomorrow.

corhere commented on libnetwork/ipams/defaultipam/allocator.go:
```diff
@@ -98,19 +90,23 @@ func NewAllocator(lcAs, glAs []*net.IPNet) (*Allocator, error) {
 
-		if p.Addr().Is4() {
-			v4 = append(v4, p)
+		n.Base = netip.PrefixFrom(n.Base.Addr().Unmap(), n.Base.Bits()).Masked()
```
```suggestion
		n.Base, _ = n.Base.Addr().Unmap().Prefix(n.Base.Bits())
```

corhere reviewed (commented):
The new allocator should work fine with Swarm. The CNM network allocator replays allocations as static assignments if there is an existing allocation in the Swarm state. https://github.com/moby/moby/blob/4554d871d70bd055a0faa3ed595419ec4dc50406/libnetwork/cnmallocator/networkallocator.go#L866-L873

corhere commented on libnetwork/ipams/defaultipam/structures_test.go:
```diff
@@ -0,0 +1,32 @@
+	"gotest.tools/v3/assert"
+)
+
+func TestDoubleCursor(t *testing.T) {
```
```suggestion
func TestMergeIter(t *testing.T) {
```

corhere commented on libnetwork/netutils/utils_linux.go:
```diff
@@ -5,34 +5,98 @@
+// Only IPv4 prefixes are returned as there's no such thing as on-link
+// routes for IPv6.
+func queryOnLinkRoutes() []netip.Prefix {
+	family := netlink.FAMILY_V4
+
+	routes, err := ns.NlHandle().RouteList(nil, family)
```
```suggestion
	routes, err := ns.NlHandle().RouteList(nil, netlink.FAMILY_V4)
```

corhere commented on libnetwork/ipams/defaultipam/address_space.go:
```diff
@@ -2,36 +2,64 @@ package defaultipam
+
+	if len(predefined) > j {
+		j++
+	}
```
Any nonzero pointers in the backing array between `predefined[j:len(predefined)]` are considered alive references for GC purposes, even after reslicing. The tail of the slice needs to be zeroed out to avoid a memory leak.

```suggestion
	}
	clear(predefined[j:len(predefined)])
```

corhere commented on libnetwork/drivers/bridge/bridge_linux_test.go:
```diff
@@ -194,17 +197,21 @@ func compareBindings(a, b []types.PortBinding) bool {
 	return true
 }
 
+var a, _ = defaultipam.NewAllocator(ipamutils.GetLocalScopeDefaultNetworks(), nil)
```
Why is this allocator package-scoped? It doesn't look like `getIPv4Data` used to have any (local) state which carried between tests, so why change it? None of the allocations are released; is there not a risk of running out?
